core/chains/terra/monitor/balance.go

```diff
@@ -0,0 +1,138 @@
+package monitor
+
+import (
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/smartcontractkit/chainlink-terra/pkg/terra"
+	"github.com/smartcontractkit/chainlink-terra/pkg/terra/client"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services"
+	"github.com/smartcontractkit/chainlink/core/services/keystore"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+func NewBalanceMonitor(chainID string, cfg terra.Config, lggr logger.Logger, ks keystore.Terra, readerFn func(string) (client.Reader, error)) services.Service {
+	return &balanceMonitor{
+		chainID:   chainID,
+		cfg:       cfg,
+		lggr:      lggr.Named("BalanceMonitor"),
+		ks:        ks,
+		newReader: readerFn,
+		stop:      make(chan struct{}),
+		done:      make(chan struct{}),
+	}
+}
+
+type balanceMonitor struct {
+	utils.StartStopOnce
+	chainID   string
+	cfg       terra.Config
+	lggr      logger.Logger
+	ks        keystore.Terra
+	newReader func(string) (client.Reader, error)
+
+	reader client.Reader
+
+	stop, done chan struct{}
+}
+
+func (b *balanceMonitor) Start() error {
+	return b.StartOnce("TerraBalanceMonitor", func() error {
+		go b.monitor()
+		return nil
+	})
+}
+
+func (b *balanceMonitor) Close() error {
+	return b.StopOnce("TerraBalanceMonitor", func() error {
+		close(b.stop)
+		<-b.done
+		return nil
+	})
+}
+
+var promTerraBalance = promauto.NewGaugeVec(
+	prometheus.GaugeOpts{Name: "terra_balance", Help: "Terra account balances"},
+	[]string{"account", "terraChainID", "denomination"},
+)
+
+func (b *balanceMonitor) updateProm(acc sdk.AccAddress, bal *sdk.DecCoin) {
+	balF, err := bal.Amount.Float64()
+	if err != nil {
+		b.lggr.Errorw("Failed to convert balance to float", "err", err)
+	}
+	promTerraBalance.WithLabelValues(acc.String(), b.chainID, bal.GetDenom()).Set(balF)
+}
+
+func (b *balanceMonitor) monitor() {
+	defer close(b.done)
+
+	tick := time.After(utils.WithJitter(b.cfg.BlockRate()))
+
+	for {
+		select {
+		case <-b.stop:
+			return
+		case <-tick:
+			b.updateBalances()
+			tick = time.After(utils.WithJitter(b.cfg.BlockRate()))
+		}
+	}
+}
+
+// getReader returns the cached client.Reader, or creates a new one if nil.
+func (b *balanceMonitor) getReader() (client.Reader, error) {
+	if b.reader == nil {
+		var err error
+		b.reader, err = b.newReader("")
+		if err != nil {
+			return nil, err
+		}
+	}
+	return b.reader, nil
+}
+func (b *balanceMonitor) updateBalances() {
+	keys, err := b.ks.GetAll()
+	if err != nil {
+		b.lggr.Errorw("Failed to get keys", "err", err)
+		return
+	}
+	if len(keys) == 0 {
+		return
+	}
+	r, err := b.getReader()
+	if err != nil {
+		b.lggr.Errorw("Failed to get client", "err", err)
+		return
+	}
+	var gotBals int
+	for _, k := range keys {
+		select {
```

Comment:
so with this the worst case shutdown time is the length of a r.Balance call right? worth a comment